Build fixes

asomers opened the pull request.

asomers committed:
Fix the build with newer compilers

syntex_syntax fails to build with recent compilers.  Fix the build by
updating bindgen to 0.31.0, which no longer uses syntex.

asomers committed:
Fix the build of sample-scsi

It was failing due to a borrow-checker error

vthriller commented:
Whoa, that new flat namespace looks wild, and I'm especially not happy about `_ => panic!()` part that this change caused. Do recent versions of bindgen also map C enum variants into individual top-level constants like `xpt_opcode_XPT_SCSI_IO`?

I can't remember what version of Rust allowed that elision of `.as_ref()`/`.as_mut()` first, but my guess is that was after 1.26 mentioned in the README. (However I'm not sure if it's worth mentioning minimum compiler version these days given changes that land in recent releases of the compiler. I guess that information would only be relevant to people that want to avoid rustup while running legacy systems like FreeBSD-10, for which only rust-1.30 is available as pkg.)

Also, this patch mixes tab and space indentation. (I guess I should provide some rustfmt config in addition to .editorconfig.)

asomers commented:
> Whoa, that new flat namespace looks wild, and I'm especially not happy about `_ => panic!()` part that this change caused. Do recent versions of bindgen also map C enum variants into individual top-level constants like `xpt_opcode_XPT_SCSI_IO`?

Yes, that's exactly what bindgen does now.  See https://github.com/rust-lang/rust-bindgen/blob/master/CHANGELOG.md#0310 .  The `panic!`, or something like it, is also now necessary, because there's no guarantee that the mapped variable has an allowed value.

> 
> I can't remember what version of Rust allowed that elision of `.as_ref()`/`.as_mut()` first, but my guess is that was after 1.26 mentioned in the README. (However I'm not sure if it's worth mentioning minimum compiler version these days given changes that land in recent releases of the compiler. I guess that information would only be relevant to people that want to avoid rustup while running legacy systems like FreeBSD-10, for which only rust-1.30 is available as pkg.)

It's not that `.as_ref()`/`.as_mut()`  are elided, it's that they aren't even allowed now.  I think it's related to bindgen's enum change.

> 
> Also, this patch mixes tab and space indentation. (I guess I should provide some rustfmt config in addition to .editorconfig.)

Aw, crap.  That must be my fault for not using my normal development system.  I'll fix it.

vthriller commented:
> Yes, that's exactly what bindgen does now. See https://github.com/rust-lang/rust-bindgen/blob/master/CHANGELOG.md#0310 . The `panic!`, or something like it, is also now necessary, because there's no guarantee that the mapped variable has an allowed value.

Thanks for the link! Now that makes perfect sense. I guess I'll let internal mod be that ugly (it's binding to an external C library, after all). I'll even let `panic!` given the overall hacky state of the library (there's already plenty of `panic!`s to TODO out) and my desire to back-port this patch into non-breaking 0.10.2.

> It's not that `.as_ref()`/`.as_mut()` are elided, it's that they aren't even allowed now. I think it's related to bindgen's enum change.

I doubt enum change is the reason, as these changes mostly affect CCB union members.

…and indeed [bindgen changelog mentions switch to Rust unions](https://github.com/rust-lang/rust-bindgen/blob/master/CHANGELOG.md#0300). So that resolves my concerns about minimum Rust version.

vthriller commented on src/cam/ccb.rs:
```diff
@@ -22,34 +22,25 @@ impl CCB {
-			for i in start..end {
-				(*ccb).bindgen_union_field[i] = 0;
-			}
-		}
```
I'm not sure it's a good idea to leave CCB un-zeroed. FreeBSD's camcontrol keeps `CCB_CLEAR_ALL_EXCEPT_HDR()` it with (almost?) every allocation ([attrib.c](https://github.com/freebsd/freebsd/blob/f55899f5a9afdebbf970ff95520093f7a780b549/sbin/camcontrol/attrib.c#L134-L141), [fwdownload.c](https://github.com/freebsd/freebsd/blob/bb0c984706c2c72ef11729bd8005eb3d108099b2/sbin/camcontrol/fwdownload.c#L462-L469), [mpt_cam.c](https://github.com/freebsd/freebsd/blob/1d6e4247415d264485ee94b59fdbc12e0c566fd0/usr.sbin/mptutil/mpt_cam.c#L241-L246) to name a few), and [smartmontools outright `bzero()`es it as well](https://github.com/mirror/smartmontools/blob/8e64dd8fef32b00cbe5fd25ad188dce4bb9f52bc/os_freebsd.cpp#L1042-L1068).

asomers replied:
Hm, let's see.  In camcontrol there are 55 instances of `cam_getccb`.  I see 33 that clear the whole thing and 15 that don't.  So it could go either way.  The problem is that `bindgen_uion_field` no longer exists.  I'll look for another way to clear it.

asomers replied:
Ok, I restored the old behavior.  Once you're satisfied, I'll squash my commits before you merge.

vthriller commented on sample-scsi/src/main.rs:
```diff
@@ -122,21 +122,23 @@ fn main() {
-				0x2f => print!("{:#?}\n", pages.informational_exceptions()),
-				_ => (),
-			}
+	let mut pages = SCSIPages::new(&dev).unwrap();
+	let supported_pages = pages.supported_pages()
+	    .iter()
+	    .cloned()
+	    .collect::<Vec<_>>();
+	for p in supported_pages {
+		if p == 00 { continue; }
+
+		print!("=== [{:02x}] {} ===\n", p, page_name(p));
+		match p {
+			// already in cli
+			0x02...0x06 | 0x0d | 0x0e => (),
+			0x10 => print!("{:#?}\n", pages.self_test_results()),
+			0x2f => print!("{:#?}\n", pages.informational_exceptions()),
+			_ => (),
```
Out of curiosity: may I ask what prompted you to do this seemingly irrelevant change?

asomers replied:
The borrow checker.  It didn't like the old code, because `pages` was borrowed read-only at line 126, and also borrowed read-write at lines 134 & 135 while the read-only borrow was still active.  The new method `collect`s the supported pages so there won't be a borrow conflict.

vthriller replied:
Oh, I see. It seems it's been a looong while since I last tried building this subcrate…

asomers committed:
Fix a segfault in ATADevice::ata_platform_do on FreeBSD

This code was previously stack-allocating a CCB, but the Drop impl
called cam_freeccb anyway.  By chance, that worked sometimes.  Other
times, it segfaulted.  Fix it by using the standard CCB allocator.

vthriller commented:
Nice catch in 753ee3323e694de0efe896c2ffbe3fd0f2460063!

> Once you're satisfied, I'll squash my commits before you merge.

LGTM, you can go ahead.

asomers commented:
All ready to merge.

vthriller commented:
Thanks for taking your time!

I'll wait until Monday before releasing 0.10.2 with commits backported onto 0.10 branch.
